synapse/storage/databases/main/relations.py

```diff
@@ -295,60 +332,96 @@ def _get_aggregation_groups_for_event_txn(
             "get_aggregation_groups_for_event", _get_aggregation_groups_for_event_txn
         )
 
-    @cached()
-    async def get_applicable_edit(
-        self, event_id: str, room_id: str
-    ) -> Optional[EventBase]:
+    async def _get_applicable_edits(
+        self, event_ids: Iterable[str]
+    ) -> Dict[str, EventBase]:
         """Get the most recent edit (if any) that has happened for the given
-        event.
+        events.
 
         Correctly handles checking whether edits were allowed to happen.
 
         Args:
-            event_id: The original event ID
-            room_id: The original event's room ID
+            event_ids: The original event IDs
 
         Returns:
-            The most recent edit, if any.
+            A map of the most recent edit for each event. A missing event implies
+            there is no edits.
         """
 
+        # A map of the original event IDs to the edit events.
+        edits_by_original = {}
+
+        # Check if an edit for this event is currently cached.
+        event_ids_to_check = []
+        for event_id in event_ids:
+            if event_id not in self.get_applicable_edit:
+                event_ids_to_check.append(event_id)
+            else:
+                edit_event = self.get_applicable_edit[event_id]
+                if edit_event:
+                    edits_by_original[event_id] = edit_event
+
+        # If all events were cached, all done.
+        if not event_ids_to_check:
+            return edits_by_original
+
         # We only allow edits for `m.room.message` events that have the same sender
         # and event type. We can't assert these things during regular event auth so
         # we have to do the checks post hoc.
 
         # Fetches latest edit that has the same type and sender as the
         # original, and is an `m.room.message`.
+        #
+        # TODO Should this ensure it does not return results for state events / redacted events?
```

Comment:
These checks were done by `_get_bundled_aggregation_for_event` for the original event, which this replaces. So I think we should do them? I wasn't sure how to check them though.

Reply:
Sounds like we do want them.

Some kind of anti-join on the `redactions` and `state_events` tables? Not sure.

Reply:
That was my initial thought too, thanks!

Reply:
We have what we need in Python (we already have an `EventBase`) so easiest to do it there, I think. See 813218006b5f752f298ed0b1afac0aa799655055.